provider/pkg/schema/jsschema_util.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2016-2021, Pulumi Corporation.
+
+package schema
+
+import (
+	jsschema "github.com/lestrrat-go/jsschema"
+)
+
+// FlattenJSSchema recursively flattens a schema containing AnyOf or OneOf into a list of schemas
+// Note: this only performs a shallow copy, don't use in situations where schema fields may be mutated.
+// TODO: this should probably handle AllOf and Not as well.
```

Comment:
Every (new) TODO in our code should reference an issue that explains how we will remove it.